finality-aleph/src/testing/proposal.rs

```diff
@@ -0,0 +1,270 @@
+        &mut aux_cip,
+        &proposal,
+        Pending(TopBlockImportedButIncorrectBranch),
+    );
+
+    chain_builder.finalize_block(&blocks[1].header.hash());
+    verify_proposal_status(
+        &mut cached_cip,
+        &mut aux_cip,
+        &proposal,
+        Pending(TopBlockImportedButIncorrectBranch),
+    );
+
+    chain_builder.finalize_block(&blocks[10].header.hash());
+    verify_proposal_status(&mut cached_cip, &mut aux_cip, &proposal, Ignore);
+}
+
+#[tokio::test]
+async fn pending_top_block_handled_correctly() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_branch_upon(&chain_builder.genesis_hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..len].to_vec();
+        let proposal = proposal_from_blocks(blocks_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Pending(PendingTopBlock),
+        );
+    }
+    chain_builder.import_branch(blocks.clone(), false).await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..len].to_vec();
+        let proposal = proposal_from_blocks(blocks_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Finalize(proposal.top_block()),
+        );
+    }
+}
+
+#[tokio::test]
+async fn hopeless_forks_handled_correctly() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_and_import_branch_upon(
+            &chain_builder.genesis_hash(),
+            MAX_DATA_BRANCH_LEN * 10,
+            false,
+        )
+        .await;
+
+    let forks = chain_builder
+        .build_branch_upon(&blocks[2].header.hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let fork_branch = forks[0..len].to_vec();
+        let proposal = proposal_from_blocks(fork_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Pending(PendingTopBlock),
+        );
+    }
+
+    chain_builder.finalize_block(&blocks[2].header.hash());
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let fork_branch = forks[0..len].to_vec();
+        let proposal = proposal_from_blocks(fork_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Pending(PendingTopBlock),
+        );
+    }
+
+    chain_builder.finalize_block(&blocks[3].header.hash());
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let fork_branch = forks[0..len].to_vec();
+        let proposal = proposal_from_blocks(fork_branch);
+        verify_proposal_status(&mut cached_cip, &mut aux_cip, &proposal, Ignore);
```

Comment:
Ok, done. To make this cleaner I have also added two options for `Ignore` the status to differentiate between hopeless fork and simply being "too low" overall.

Reply:
I'm not convinced this distinction is useful, since we don't care about it anywhere in the code (outside of maybe tests and logging, the latter could be provided in another way). I would rather keep `Ignore` without a `reason` inside because of this.

Reply:
I'm ok with removing it, but I don't understand what is the harm of having this additional piece of information. How does it make the code worse?

The benefits on the other hand are clear: more precise test and better info when debugging.

Reply:
For logging you can add log lines in the ifs that return the variants now.

Tests are not supposed to be precise in the sense of testing every `if` in the code they are testing, but only in testing all the properties we are actually interested in in code that uses it. Since in actual code we only act differently for `Ignore` vs anything else, disregarding the `reason`, we shouldn't care about it in the tests either. In this specific case we could theoretically switch the `if`s around and the code would still be correct (telling us to ignore the provided proposals and even giving correct reasons for that), but the tests would start failing.

Reply:
Ok, I removed the reason and only left one log for catching hopeless forks.